# XML, HTML, and Web Scraping

JSON and XML are two different ways to represent hierarchical data. Which one is better? There are lots of articles online which discuss similarities and differences between JSON and XML and their advantages and disadvantages. Both formats are still in current usage, so it is good to be familiar with both. However, JSON is more common, so we'll focus on working with JSON representations of hierarchical data.

The reading covered an example of using Beautiful Soup to parse XML. Rather than doing another example XML now, we'll skip straight to scraping HTML from a webpage. Both HTML and XML can be parsed in a similar way with Beautiful Soup.

In [ ]:
import pandas as pd
!pip install bs4
!pip install requests

## Scraping an HTML table with Beautiful Soup

Open the URL https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population and scroll down until you see a table of the cities in the U.S. with population over 100,000 (as of Jul 1, 2022). We'll use Beautiful Soup to scrape information from this table.

Read in the HTML from the ULR using the `requests` library.

In [ ]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

[                    Population tables of U.S. cities
 0  The skyline of New York City, the most populou...
 1                                             Cities
 2  Population AreaDensityEthnic identityForeign-b...
 3                                        Urban areas
 4             Populous cities and metropolitan areas
 5                                 Metropolitan areas
 6  184 combined statistical areas935 core-based s...
 7                                        Megaregions
 8  See related population listsNorth American met...
 9                                                vte,
     0                               1
 0 NaN                   State capital
 1 NaN              State largest city
 2 NaN  State capital and largest city
 3 NaN                 Federal capital,
             City  ST 2023 estimate 2020 census  Change 2020 land area          \
             City  ST 2023 estimate 2020 census  Change            mi2     km2   
 0    New York[c]  NY       8258035     88041

Use Beautiful Soup to parse this string into a tree called `soup`

---



In [ ]:
import requests
response = requests.get("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

In [ ]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(response.content, "html.parser")



<!DOCTYPE html>

<html class="client-nojs vector-feature-language-in-header-enabled vector-feature-language-in-main-page-header-disabled vector-feature-sticky-header-disabled vector-feature-page-tools-pinned-disabled vector-feature-toc-pinned-clientpref-1 vector-feature-main-menu-pinned-disabled vector-feature-limited-width-clientpref-1 vector-feature-limited-width-content-enabled vector-feature-custom-font-size-clientpref-1 vector-feature-appearance-pinned-clientpref-1 vector-feature-night-mode-enabled skin-theme-clientpref-day vector-toc-available" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of United States cities by population - Wikipedia</title>
<script>(function(){var className="client-js vector-feature-language-in-header-enabled vector-feature-language-in-main-page-header-disabled vector-feature-sticky-header-disabled vector-feature-page-tools-pinned-disabled vector-feature-toc-pinned-clientpref-1 vector-feature-main-menu-pinned-disabled vector-feature-limite

In [ ]:
# YOUR CODE HERE

To find an HTML tag corresponding to a specific element on a webpage, right-click on it and choose "Inspect element". Go to the cities table Wikipedia page and do this now.

You should find that the cities table on the Wikipedia page corresponds to the element

```
<table class="wikitable sortable jquery-tablesorter" style="text-align:center">
```

There are many `<table>` tags on the page.

In [ ]:
len(soup.find_all("table"))

10

We can use attributes like `class=` and `style=` to narrow down the list.

In [ ]:
len(soup.find_all("table",
                  attrs={
                      "class": "wikitable"}
                  ))

7

At this point, you can manually inspect the tables on the webpage to find that the one we want is the first one (see `[0]` below). We'll store this as `table`.

In [ ]:
table = soup.find_all("table",
                  attrs={
                      "class": "wikitable"}
                  )[1]
table

<table class="sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center" style="text-align:right">
<tbody><tr>
<th rowspan="2">City
</th>
<th rowspan="2"><abbr title="State or district">ST</abbr>
</th>
<th rowspan="2">2023<br/>estimate
</th>
<th rowspan="2">2020<br/>census
</th>
<th rowspan="2">Change
</th>
<th colspan="2">2020 land area
</th>
<th colspan="2">2020 density
</th>
<th rowspan="2">Location
</th></tr>
<tr class="static-row-header">
<th>mi<sup>2</sup>
</th>
<th>km<sup>2</sup>
</th>
<th>/ mi<sup>2</sup>
</th>
<th>/ km<sup>2</sup>
</th></tr>
<tr>
<td style="background-color:#cfecec"><i><a href="/wiki/New_York_City" title="New York City">New York</a></i><sup class="reference" id="cite_ref-5"><a href="#cite_note-5"><span class="cite-bracket">[</span>c<span class="cite-bracket">]</span></a></sup>
</td>
<td><a href="/wiki/New_York_(state)" title="New York (state)">NY</a>
</td>
<td style="text-align:right;">8,258,035
</td>
<td style="text-align:right;

**Now you will write code to scrape the information in `table` to create a Pandas data frame with one row for each city and columns for: city, state, population (2022 estimate), and 2020 land area (sq mi).** Refer to the Notes/suggestions below as you write your code. A few Hints are provided further down, but try coding first before looking at the hints.

Notes/suggestions:

- Use as a guide the code from the reading that produced the data frame of Statistics faculty
- Inspect the page source as you write your code
- You will need to write a loop to get the information for all cities, but you might want to try just scraping the info for New York first
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `.get_text(strip = True)` instead of `.text`
- Don't forget to convert to a Pandas Data Frame; it should have 333 rows and 4 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it --- e.g., what is the population density for all cities in CA? --- then you would need to clean the data first (to clean strings and convert to quantitative). (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)

In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [ ]:
city = table.find_all("tr")[2]
cells = city.find_all("td")
cells[5].text

'300.5\n'

In [ ]:
# initialize an empty list
rows = []

# iterate over all rows in the faculty table
for city in table.find_all("tr")[2:]:

    # Get all the cells (<td>) in the row.
    cells = city.find_all("td")

    # The information we need is the text between tags.

    # Find the the name of the faculty in cell[0]
    # which for most faculty is contained in the <strong> tag
    City = cells[0]      #.find_all("strong") or cells[0]
    City = City.text.replace("\n", "")
    if City.endswith("]"):
      City = City[:-3]
    #-1:]""
    #City = City.text
#if city.ends with: city=city.rstrip -3:
    # Find the office of the faculty in cell[1]
    # which for most faculty is contained in the <a> tag
    population = cells[2]      #.find_all("a") or cells[2]
    population = population.text.replace("\n", "")

    # Find the email of the faculty in cell[3]
    # which for most faculty is contained in the <a> tag
    land_density = cells[5]      #.find_all("a") or cells[5]
    land_density = land_density.text.replace("\n", "")

    # Append this data.
    rows.append({
        "City": City,
        "23_Population ": population,
        "20_land_area": land_density
    })

In [ ]:
df = pd.DataFrame(rows)
df

City 23_Population  20_land_area
0       New York      8,258,035        300.5
1    Los Angeles      3,820,914        469.5
2        Chicago      2,664,452        227.7
3        Houston      2,314,157        640.4
4        Phoenix      1,650,070        518.0
..           ...            ...          ...
331         Yuma        100,858        120.7
332  New Bedford        100,695         20.0
333      Suffolk        100,659        399.2
334     Hesperia        100,633         72.7
335    Davenport        100,354         63.8

[336 rows x 3 columns]

Hints:

- Each city is a row in the table; find all the `<tr>` tags to find all the cities

- Look for the `<td>` tag to see table entries within a row
- The rank column is represented by `<th>` tags, rather than `<td>` tags. So within a row, the first (that is, `[0]`) `<td>` tag corresponds to the city name.

## Aside: Scraping an HTML table with Pandas



The Pandas command `read_html` can be used to scrape information from an HTML table on a webpage.

We can call `read_html` on the URL.

In [ ]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

[                    Population tables of U.S. cities
 0  The skyline of New York City, the most populou...
 1                                             Cities
 2  Population AreaDensityEthnic identityForeign-b...
 3                                        Urban areas
 4             Populous cities and metropolitan areas
 5                                 Metropolitan areas
 6  184 combined statistical areas935 core-based s...
 7                                        Megaregions
 8  See related population listsNorth American met...
 9                                                vte,
     0                               1
 0 NaN                   State capital
 1 NaN              State largest city
 2 NaN  State capital and largest city
 3 NaN                 Federal capital,
             City  ST 2023 estimate 2020 census  Change 2020 land area          \
             City  ST 2023 estimate 2020 census  Change            mi2     km2   
 0    New York[c]  NY       8258035     88041

However, this scrapes all the tables on the webpage, not just the one we want. As with Beautiful Soup, we can narrow the search by specifying the table attributes.

In [ ]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs={'class': 'wikitable', 'style': 'text-align:right'})


[            City  ST 2023 estimate 2020 census  Change 2020 land area          \
             City  ST 2023 estimate 2020 census  Change            mi2     km2   
 0    New York[c]  NY       8258035     8804190  −6.20%          300.5   778.3   
 1    Los Angeles  CA       3820914     3898747  −2.00%          469.5  1216.0   
 2        Chicago  IL       2664452     2746388  −2.98%          227.7   589.7   
 3        Houston  TX       2314157     2304580  +0.42%          640.4  1658.6   
 4        Phoenix  AZ       1650070     1608139  +2.61%          518.0  1341.6   
 ..           ...  ..           ...         ...     ...            ...     ...   
 331         Yuma  AZ        100858       95548  +5.56%          120.7   312.6   
 332  New Bedford  MA        100695      101079  −0.38%           20.0    51.8   
 333   Suffolk[l]  VA        100659       94324  +6.72%          399.2  1033.9   
 334     Hesperia  CA        100633       99818  +0.82%           72.7   188.3   
 335    Davenpor

This still returns 3 tables. As we remarked above, the table that we want is the first one (see `[0]` below).

In [ ]:
df = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs={'class': 'wikitable', 'style': 'text-align:right'})[0]
df

City  ST 2023 estimate 2020 census  Change 2020 land area          \
            City  ST 2023 estimate 2020 census  Change            mi2     km2   
0    New York[c]  NY       8258035     8804190  −6.20%          300.5   778.3   
1    Los Angeles  CA       3820914     3898747  −2.00%          469.5  1216.0   
2        Chicago  IL       2664452     2746388  −2.98%          227.7   589.7   
3        Houston  TX       2314157     2304580  +0.42%          640.4  1658.6   
4        Phoenix  AZ       1650070     1608139  +2.61%          518.0  1341.6   
..           ...  ..           ...         ...     ...            ...     ...   
331         Yuma  AZ        100858       95548  +5.56%          120.7   312.6   
332  New Bedford  MA        100695      101079  −0.38%           20.0    51.8   
333   Suffolk[l]  VA        100659       94324  +6.72%          399.2  1033.9   
334     Hesperia  CA        100633       99818  +0.82%           72.7   188.3   
335    Davenport  IA        100354      101724  −1.35%           63.8   165.2   

    2020 density                                      Location  
           / mi2  / km2                               Location  
0          29298  11312    40°40′N 73°56′W / 40.66°N 73.94°W  
1           8304   3206  34°01′N 118°25′W / 34.02°N 118.41°W  
2          12061   4657    41°50′N 87°41′W / 41.84°N 87.68°W  
3           3599   1390    29°47′N 95°23′W / 29.79°N 95.39°W  
4           3105   1199  33°34′N 112°05′W / 33.57°N 112.09°W  
..           ...    ...                                    ...  
331          792    306  32°31′N 114°31′W / 32.52°N 114.52°W  
332         5054   1951    41°40′N 70°56′W / 41.66°N 70.94°W  
333          236     91    36°42′N 76°38′W / 36.70°N 76.63°W  
334         1373    530  34°24′N 117°19′W / 34.40°N 117.32°W  
335         1594    615    41°34′N 90°36′W / 41.56°N 90.60°W  

[336 rows x 10 columns]

Wait, that seemed much easier than using Beautiful Soup, and it returned a data frame, and we even got for free some formatting like removing the commas from the population! Why didn't we just use `read_html` in the first place? It's true the `read_html` works well when scraping information from an HTML *table*. Unfortunately, you often want to scrape information from a webpage that isn't conveniently stored in an HTML table, in which case `read_html` won't work. (It only searches for `<table>`, `<th>`, `<tr>`, and `<td>` tags, but there are many other HTML tags.) Though Beautiful Soup is not as simple as `read_html`, it is more flexible and thus more widely applicable.

## Scraping information that is NOT in a `<table>` with Beautiful Soup

The Cal Poly course catalog http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory contains a list of courses offered by the Statistics department. **You will scrape this website to obtain a Pandas data frame with one row for each DATA or STAT course and two columns: course name and number (e.g, DATA 301. Introduction to Data Science) and term typically offered (e.g., Term Typically Offered: F, W, SP).**

Note: Pandas `read_html` is not help here since the courses are not stored in a `<table>.`

In [ ]:
pd.read_html("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")

[                                        Program name   Program type
 0                              Actuarial Preparation          Minor
 1  Cross Disciplinary Studies Minor in Bioinforma...          Minor
 2   Cross Disciplinary Studies Minor in Data Science          Minor
 3                                         Statistics  BS, MS, Minor]


Notes/suggestions:


- Inspect the page source as you write your code
- The courses are not stored in a `<table>`. How are they stored?
- You will need to write a loop to get the information for all courses, but you might want to try just scraping the info for DATA 100 first
- What kind of tag is the course name stored in? What is the `class` of the tag?
- What kind of tag is the quarter(s) the course is offered stored in? What is the `class` of the tag? Is this the only tag of this type with the class? How will you get the one you want?
- You don't have to remove the number of units (e.g., 4 units) from the course name and number, but you can try it if you want
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `get_text(strip = True)` instead of `text`
- Don't forget to convert to a Pandas Data Frame; it should have 74 rows and 2 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it then you might need to clean the data first. (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)



In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [ ]:
response1 = requests.get("https://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")
soup1 = BeautifulSoup(response1.content, "html.parser")



In [ ]:
table2 = soup1.find_all("div", {"class": "courseblock"})
table2

[<div class="courseblock">
 <p class="courseblocktitle"><strong>DATA 100. Data Science for All I.
 <span class="courseblockhours">4 units
 </span></strong></p><div class="noindent courseextendedwrap">
 <p class="noindent">Term Typically Offered: F, W, SP</p><p class="noindent">2020-21 or later catalog: GE Area B4</p><p class="noindent">2019-20 or earlier catalog: GE Area B4</p><p>Prerequisite: <a class="bubblelink code" href="/search/?P=MATH%20115" onclick="return showCourse(this, 'MATH 115');" title="MATH 115">MATH 115</a>, <a class="bubblelink code" href="/search/?P=MATH%20116" onclick="return showCourse(this, 'MATH 116');" title="MATH 116">MATH 116</a>, <a class="bubblelink code" href="/search/?P=MATH%20118" onclick="return showCourse(this, 'MATH 118');" title="MATH 118">MATH 118</a>, or Appropriate Math Placement Level.</p></div>
 <div class="courseblockdesc">
 <p>Basic approaches for answering questions using data.  Emphasis on working with tabular data in spreadsheet software to 

In [ ]:
course_title = soup1.find('p', class_='courseblocktitle')
course_title = course_title.text.strip()
course_title

AttributeError: ResultSet object has no attribute 'text'. You're probably treating a list of elements like a single element. Did you call find_all() when you meant to call find()?

In [ ]:
course_titles = soup1.find_all('p', class_='courseblocktitle')[1]
course_titles

<p class="courseblocktitle"><strong>DATA 301. Introduction to Data Science.
<span class="courseblockhours">4 units
</span></strong></p>

In [ ]:
data= soup1.find_all("div", {"class": "courseblock"})[1]
data

<div class="courseblock">
<p class="courseblocktitle"><strong>DATA 301. Introduction to Data Science.
<span class="courseblockhours">4 units
</span></strong></p><div class="noindent courseextendedwrap">
<p class="noindent">Term Typically Offered: F, W, SP</p><p>Prerequisite: CPE/<a class="bubblelink code" href="/search/?P=CSC%20202" onclick="return showCourse(this, 'CSC 202');" title="CSC 202">CSC 202</a>; and one of the following: <a class="bubblelink code" href="/search/?P=IME%20326" onclick="return showCourse(this, 'IME 326');" title="IME 326">IME 326</a>, <a class="bubblelink code" href="/search/?P=STAT%20302" onclick="return showCourse(this, 'STAT 302');" title="STAT 302">STAT 302</a>, <a class="bubblelink code" href="/search/?P=STAT%20312" onclick="return showCourse(this, 'STAT 312');" title="STAT 312">STAT 312</a>, or <a class="bubblelink code" href="/search/?P=STAT%20313" onclick="return showCourse(this, 'STAT 313');" title="STAT 313">STAT 313</a>.</p></div>
<div class="courseb

In [ ]:
term = soup1.find('p', class_='noindent')
term = term.text.strip()
term

'Term Typically Offered: F, W, SP'

In [ ]:
course = soup1.find_all("div", {"class": "courseblock"})[1]
#course
cells= course.find_all('p')
cells[1].text

'Term Typically Offered: F, W, SP'

In [ ]:
cells = course.find_all("div")
cells[0].text

'\nTerm Typically Offered: F, W, SPPrerequisite: CPE/CSC\xa0202; and one of the following: IME\xa0326, STAT\xa0302, STAT\xa0312, or STAT\xa0313.'

In [ ]:
# initialize an empty list
rows1 = []

# iterate over all rows in the faculty table
for Course in soup1.find_all("div", {"class": "courseblock"}):

    # Get all the cells (<td>) in the row.
    cells= Course.find_all('p')

    # The information we need is the text between tags.

    # Find the the name of the faculty in cell[0]
    # which for most faculty is contained in the <strong> tag
    course_title = cells[0]
    course_title = course_title.text

    # Find the office of the faculty in cell[1]
    # which for most faculty is contained in the <a> tag
    term = cells[1]
    term = term.text

    # Find the email of the faculty in cell[3]
    # which for most faculty is contained in the <a> tag
    #email_tag = cells[3].find("a") or cells[3]
    #email = email_tag.text

    # Append this data.
    rows1.append({
        "Course": course_title,
        "Terms_offered": term,

    })

In [ ]:
City = cells[0]      #.find_all("strong") or cells[0]
    City = City.text.replace("\n", "")
    if City.endswith("]"):
      City = City[:-3]

In [ ]:
df = pd.DataFrame(rows1)
df

Course  \
0        DATA 100. Data Science for All I.\n4 units\n   
1   DATA 301. Introduction to Data Science.\n4 uni...   
2   DATA 401. Data Science Process and Ethics.\n3 ...   
3   DATA 402. Mathematical Foundations of Data Sci...   
4   DATA 403. Data Science Projects Laboratory.\n1...   
..                                                ...   
69  STAT 551. Statistical Learning with R.\n4 units\n   
70  STAT 566. Graduate Consulting Practicum.\n2 un...   
71   STAT 570. Selected Advanced Topics.\n1-4 units\n   
72  STAT 590. Graduate Seminar in Statistics.\n1 u...   
73                     STAT 599. Thesis.\n1-4 units\n   

                       Terms_offered  
0   Term Typically Offered: F, W, SP  
1   Term Typically Offered: F, W, SP  
2          Term Typically Offered: F  
3          Term Typically Offered: F  
4          Term Typically Offered: F  
..                               ...  
69         Term Typically Offered: F  
70        Term Typically Offered: SP  
71       Term Typically Offered: TBD  
72  Term Typically Offered: F, W, SP  
73       Term Typically Offered: TBD  

[74 rows x 2 columns]

Hints:

- Each course is represented by a `<div>` with `class=courseblock`, so you can find all the courses with `soup.find_all("div", {"class": "courseblock"})`
- The course name is in a `<p>` tag with `class=courseblocktitle`, inside a `<strong>` tag. (Though I don't think we need to find the strong tag here.)
- The term typically offered is in `<p>` tag with `class=noindent`. However, there are several tags with this class; term typically offered is the first one.
- If you want to use Beautiful Soup to remove the course units (e.g., 4 units), find the `<span>` tag within the course name tag and `.extract()` this span tag